### PR TITLE
Add post processor incremental core logic and tests

### PR DESCRIPTION
**Pull request by hellosnow**

### CODE REVIEW

File: test/Microsoft.DocAsCode.Build.Engine.Tests/PostProcessors/NotSupportIncrementalPostProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Build.Engine.Tests
+{
+    using System.Collections.Immutable;
+
+    using Microsoft.DocAsCode.Plugins;
+
+    internal class NotSupportIncrementalPostProcessor : IPostProcessor
```
Comment (vicancy) [inline test/Microsoft.DocAsCode.Build.Engine.Tests/PostProcessors/NotSupportIncrementalPostProcessor.cs]: >NotSupportIncremental [](start = 19, length = 21)

NonIncremental or IncrementalNotSupported?

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs [hunk 2]
```diff
@@ -42,8 +46,150 @@ public void Handle(List<PostProcessor> postProcessors, Manifest manifest, string
...
+            }
+            TraceIntermediateInfo(outputFolder, increItems, noneIncreItems);
+            PostHandle(manifest, increItems);
+        }
+
+        #region Handle related
+
+        private void PreHandle(Manifest manifest, string outputFolder, List<ManifestItem> increItems, List<ManifestItem> noneIncreItems)
+        {
+            if (_increContext.CanIncrementalPostProcess)
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs]: >CanIncrementalPostProcess [](start = 30, length = 25)

naming: CanIncremental is enough?

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs [hunk 3]
```diff
@@ -42,8 +46,150 @@ public void Handle(List<PostProcessor> postProcessors, Manifest manifest, string
...
-            // TODO: incremental things
-            _innerHandler.Handle(postProcessors, manifest, outputFolder);
+            // TODO: add logs
+            // TODO: replay warning
+            var increItems = manifest.Files.Where(i => i.IsIncremental).ToList();
+            var noneIncreItems = manifest.Files.Where(i => !i.IsIncremental).ToList();
+
+            PreHandle(manifest, outputFolder, increItems, noneIncreItems);
+            {
+                CheckNoIncrementalItems(manifest);
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs]: >CheckNoIncrementalItems [](start = 16, length = 23)

no need? as the logic before guarantees all items `IsIncremental == false`?

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs [hunk 4]
```diff
@@ -42,8 +46,150 @@ public void Handle(List<PostProcessor> postProcessors, Manifest manifest, string
...
+            return from item in items
+                   from output in item.OutputFiles.Values
+                select output.RelativePath;
+        }
+
+        private static void CheckNoIncrementalItems(Manifest manifest)
+        {
+            if (manifest.Files.Any(i => i.IsIncremental))
+            {
+                throw new DocfxException("Manifest items are not supposed to have any incremental items.");
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs]: >"Manifest items are not supposed to have any incremental items." [](start = 41, length = 64)

This message is not quite accurate>? `Handle` does not allow manifest items to be incremental, however PostHandle adds incremental items.

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs]: yes should be more specific

---
In reply to: [95292970](https://github.com/dotnet/docfx/pull/1136#discussion_r95292970) [](ancestors = 95292970)

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs (hunk 3 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs]: want to double confirm no incremental items go into inner logic...

---
In reply to: [95292681](https://github.com/dotnet/docfx/pull/1136#discussion_r95292681) [](ancestors = 95292681)

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs [hunk 5]
```diff
@@ -42,8 +46,150 @@ public void Handle(List<PostProcessor> postProcessors, Manifest manifest, string
...
+                {
+                    throw new BuildCacheException($"Last incremental post processor outputs shuold contain {outputRelPath}.");
+                }
+
+                IncrementalUtility.RetryIO(() =>
+                {
+                    // Copy last cached file to output
+                    var outputPath = Path.Combine(outputFolder, outputRelPath);
+                    var lastCachedFile = Path.Combine(_increContext.LastBaseDir, lastCachedRelPath);
+                    Directory.CreateDirectory(Path.GetDirectoryName(outputPath));
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs]: >Path.GetDirectoryName(outputPath) [](start = 46, length = 33)

can be empty?

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs [hunk 6]
```diff
@@ -42,8 +46,150 @@ public void Handle(List<PostProcessor> postProcessors, Manifest manifest, string
...
+            foreach (var outputRelPath in GetOutputRelativePaths(increItems))
+            {
+                string lastCachedRelPath;
+                if (_increContext.LastInfo == null)
+                {
+                    throw new BuildCacheException("Last incremental post processor info should not be null.");
+                }
+                if (!_increContext.LastInfo.PostProcessOutputs.TryGetValue(outputRelPath, out lastCachedRelPath))
+                {
+                    throw new BuildCacheException($"Last incremental post processor outputs shuold contain {outputRelPath}.");
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs]: >shuold [](start = 92, length = 6)

typo

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs [hunk 7]
```diff
@@ -42,8 +46,152 @@ public void Handle(List<PostProcessor> postProcessors, Manifest manifest, string
...
 
-            // TODO: incremental things
-            _innerHandler.Handle(postProcessors, manifest, outputFolder);
+            // TODO: add logs
+            // TODO: replay warning
+            var increItems = manifest.Files.Where(i => i.IsIncremental).ToList();
+            var noneIncreItems = manifest.Files.Where(i => !i.IsIncremental).ToList();
+
+            PreHandle(manifest, outputFolder, increItems, noneIncreItems);
+            {
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs]: >{ [](start = 12, length = 1)

why put in `{` and `}`?

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs [hunk 8]
```diff
@@ -42,8 +46,152 @@ public void Handle(List<PostProcessor> postProcessors, Manifest manifest, string
...
+                    throw new BuildCacheException($"Last incremental post processor outputs should contain {outputRelPath}.");
+                }
+
+                IncrementalUtility.RetryIO(() =>
+                {
+                    var lastCachedFile = Path.Combine(_increContext.LastBaseDir, lastCachedRelPath);
+                    var currentCachedFileName = IncrementalUtility.GetRandomEntry(_increContext.CurrentBaseDir);
+
+                    // Copy last cached file to current cached file
+                    File.Copy(lastCachedFile, Path.Combine(_increContext.CurrentBaseDir, currentCachedFileName));
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs]: >lastCachedFile [](start = 30, length = 14)

who is going to clear last cache?

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs (hunk 7 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs]: To group inner as core logic...is optional here

---
In reply to: [95303118](https://github.com/dotnet/docfx/pull/1136#discussion_r95303118) [](ancestors = 95303118)

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs (hunk 8 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs]: in last step of Build in DocumentBuilder, refer to [https://github.com/dotnet/docfx/blob/dev/src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs#L158](https://github.com/dotnet/docfx/blob/dev/src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs#L158)

---
In reply to: [95303401](https://github.com/dotnet/docfx/pull/1136#discussion_r95303401) [](ancestors = 95303401)

**Comment by vicancy** [comment]
:shipit:

**Comment by ansyral** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/IncrementalPostProcessorsContext.cs [hunk 9]
```diff
@@ -125,12 +126,13 @@ private bool GetCanIncremental()
...
                 if (!currentPostProcessorInfo.Equals(lastPostProcessorInfo))
                 {
-                    var message = $"Cannot support incremental post processors, because post processor info changed from last {lastPostProcessorInfo.ToJsonString()} to current {currentPostProcessorInfo.ToJsonString()}.";
+                    var message = $"Cannot support incremental post processing, because post processor info changed from last {lastPostProcessorInfo.ToJsonString()} to current {currentPostProcessorInfo.ToJsonString()}.";
                     Logger.LogVerbose(message);
                     return false;
                 }
             }
 
+            Logger.LogVerbose("Can support incremental post processing.");
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/IncrementalPostProcessorsContext.cs]: `Can support` -> `Enable`?

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/IncrementalPostProcessorsContext.cs]: "enable incremental" is a switch to turn on/off the incremental post process?

---
In reply to: [95506929](https://github.com/dotnet/docfx/pull/1136#discussion_r95506929) [](ancestors = 95506929)

**Comment by vwxyzh** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/IncrementalPostProcessorsContext.cs (hunk 9 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/IncrementalPostProcessorsContext.cs]: renamed to "IsIncremental"

---
In reply to: [95507816](https://github.com/dotnet/docfx/pull/1136#discussion_r95507816) [](ancestors = 95507816,95506929)

**Pull request merged**